sql/hive/src/main/scala/org/apache/spark/sql/hive/client/package.scala

```diff
@@ -19,15 +19,27 @@ package org.apache.spark.sql.hive
 
 /** Support for interacting with different versions of the HiveMetastoreClient */
 package object client {
-  private[client] abstract class HiveVersion(val fullVersion: String, val hasBuiltinsJar: Boolean)
+  private[client] abstract class HiveVersion(
+      val fullVersion: String,
+      val extraDeps: Seq[String] = Nil,
+      val exclusions: Seq[String] = Nil)
 
   // scalastyle:off
   private[client] object hive {
-    case object v10 extends HiveVersion("0.10.0", true)
-    case object v11 extends HiveVersion("0.11.0", false)
-    case object v12 extends HiveVersion("0.12.0", false)
-    case object v13 extends HiveVersion("0.13.1", false)
+    case object v12 extends HiveVersion("0.12.0")
+    case object v13 extends HiveVersion("0.13.1")
+
+    // Hive 0.14 depends on calcite 0.9.2-incubating-SNAPSHOT which does not exist in
+    // maven central anymore, so override those with a version that exists.
+    //
+    // org.pentaho:pentaho-aggdesigner-algorithm is also nowhere to be found, so exclude
+    // it explicitly. If it's needed by the metastore client, users will have to dig it
+    // out of somewhere and use configuration to point Spark at the correct jars.
```

Comment:
By default Spark SQL will use the hive jars that are bundled with Spark, and for read deployments I expect most users will do as @vanzin suggests and set `spark.sql.hive.metastore.jars`.  This is really just for testing all the versions easily.  If someone wants to come up with a way to easily package all the various hive versions and find a way to get them on to jenkins / other places people want to run the tests we could use that instead.

@vanzin maybe we should open a JIRA for Hive though, as it would be good if we can depend on them directly eventually and the probably don't realize that is not possible.